### PR TITLE
Issue #SB-26347 fix: change the nodebb database from mongo to redis

### PR DESCRIPTION
**Pull request by VenkatKadiveti**
<table>
<td>
  <a href="https://project-sunbird.atlassian.net/browse/SB-26347" title="SB-26347" target="_blank">TaskSB-26347</a>  Discussion Forum >> change the nodebb database from mongo to redis
  </td></table>
  <br />
 

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

### Type of change

Please choose appropriate options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Enhancement (additive changes to improve performance)
- [ ] This change requires a documentation update

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes in the below checkboxes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Ran Test A
- [ ] Ran Test B

**Test Configuration**:
* Software versions:
* Hardware versions:

### Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: ansible/roles/stack-sunbird/templates/sunbird_nodebb.env [hunk 1]
```diff
@@ -1,10 +1,15 @@
 # Initialization variable for mongo
 # Ref: https://community.nodebb.org/topic/4325/set-up-nodebb-via-environment-variables
-database=mongo
+database=redis
+# redis database number
+redis_database={{nodebb_redis_database|default('0')}}
+# declearing js heap memory size
+nodebb_heap_size={{nodebb_heap_size|default('2048')}}
```
Comment (keshavprasadms) [inline ansible/roles/stack-sunbird/templates/sunbird_nodebb.env:7]: the left hand side of this should the variable name used for nodejs heap memory - `NODE_OPTIONS`
Or we can also add this value as a env in deployment manifest as
```
env:
  name: NODE_OPTIONS
  value: XXX
```

File: ansible/roles/stack-sunbird/templates/sunbird_nodebb.env [hunk 2]
```diff
@@ -1,10 +1,15 @@
 # Initialization variable for mongo
 # Ref: https://community.nodebb.org/topic/4325/set-up-nodebb-via-environment-variables
-database=mongo
+database=redis
+# redis database number
+redis_database={{nodebb_redis_database|default('0')}}
```
Comment (keshavprasadms) [inline ansible/roles/stack-sunbird/templates/sunbird_nodebb.env:5]: Default please change to a free database in exisitng redis. We will not set default as 0. 
0 will be used only for prod.

File: ansible/roles/stack-sunbird/templates/sunbird_nodebb.env [hunk 3]
```diff
@@ -1,10 +1,15 @@
...
+# declearing js heap memory size
+nodebb_heap_size={{nodebb_heap_size|default('2048')}}
 secret={{nodebb_hash_key|default('1d57ba64-86d4-43ff-bd10-f6e9e0782899')}}
 url=http://0.0.0.0:4567/discussions/
-mongo__host={{mongo_nodebb_host}}
-mongo__database=nodebb
+redis__host={{sunbird_lp_redis_host }}
```
Comment (keshavprasadms) [inline ansible/roles/stack-sunbird/templates/sunbird_nodebb.env:10]: Please reach out to @rjshrjndrn and decide on which redis should we use. KP / DP?

File: ansible/roles/stack-sunbird/templates/sunbird_nodebb.env [hunk 4]
```diff
@@ -1,10 +1,15 @@
 # Initialization variable for mongo
 # Ref: https://community.nodebb.org/topic/4325/set-up-nodebb-via-environment-variables
-database=mongo
+database=redis
+# redis database number
+redis_database={{nodebb_redis_database|default('0')}}
+# declearing js heap memory size
+nodebb_heap_size={{nodebb_heap_size|default('2048')}}
 secret={{nodebb_hash_key|default('1d57ba64-86d4-43ff-bd10-f6e9e0782899')}}
```
Comment (rjshrjndrn) [inline ansible/roles/stack-sunbird/templates/sunbird_nodebb.env:8]: What's this variable for? and why are we using a default value for the hash key?

**Review by rjshrjndrn — APPROVED** [review]

File: ansible/roles/stack-sunbird/templates/sunbird_nodebb.env (hunk 3 above)
Comment (rjshrjndrn) [reply ansible/roles/stack-sunbird/templates/sunbird_nodebb.env:10]: KP looks good to me.

File: ansible/roles/stack-sunbird/templates/sunbird_nodebb.env (hunk 4 above)
Comment (VenkatKadiveti) [reply ansible/roles/stack-sunbird/templates/sunbird_nodebb.env:8]: This is nodebb secret key and you only added this

File: ansible/roles/stack-sunbird/templates/sunbird_nodebb.env (hunk 3 above)
Comment (VenkatKadiveti) [reply ansible/roles/stack-sunbird/templates/sunbird_nodebb.env:10]: Added username and password for redis db

File: ansible/roles/stack-sunbird/templates/sunbird_nodebb.env (hunk 1 above)
Comment (VenkatKadiveti) [reply ansible/roles/stack-sunbird/templates/sunbird_nodebb.env]: changed, you can see the latest changes now

File: ansible/roles/stack-sunbird/templates/sunbird_nodebb.env [hunk 5]
```diff
@@ -1,10 +1,17 @@
 # Initialization variable for mongo
 # Ref: https://community.nodebb.org/topic/4325/set-up-nodebb-via-environment-variables
-database=mongo
+database=redis
+# redis database number
+redis_database={{nodebb_redis_database}}
```
Comment (keshavprasadms) [inline ansible/roles/stack-sunbird/templates/sunbird_nodebb.env:5]: This should have some default value. Please reach out to KP team and ask the db numbers they are using. Then we can have a default DB number. 
Example
`redis_database={{nodebb_redis_database | default(1)}}`

Comment (VenkatKadiveti) [reply ansible/roles/stack-sunbird/templates/sunbird_nodebb.env:5]: ok

File: ansible/roles/stack-sunbird/templates/sunbird_nodebb.env [hunk 6]
```diff
@@ -1,10 +1,17 @@
...
+# declearing js heap memory size
+NODE_OPTIONS={{nodebb_heap_size|default('2048')}}
```
Comment (keshavprasadms) [inline ansible/roles/stack-sunbird/templates/sunbird_nodebb.env:7]: Please reduce default heap size to 1024. Is there a problem with 1024?
All the pods in other env's have a maximum pod memory set to 1GB. So if we increase heap to more than 1GB, pod will crash if it tries to use more heap.

File: ansible/roles/stack-sunbird/templates/sunbird_nodebb.env [hunk 7]
```diff
@@ -1,10 +1,17 @@
...
 secret={{nodebb_hash_key|default('1d57ba64-86d4-43ff-bd10-f6e9e0782899')}}
 url=http://0.0.0.0:4567/discussions/
-mongo__host={{mongo_nodebb_host}}
-mongo__database=nodebb
+redis__host={{sunbird_lp_redis_host }}
+redis__port=6379
+redis__username={{nodebb_redis_username}}
```
Comment (keshavprasadms) [inline ansible/roles/stack-sunbird/templates/sunbird_nodebb.env:12]: Please add `default('')` as empty values for this. Our redis instances dont have username and password, only azure redis has login data. Can we have empty value for these? Will it work?

File: ansible/roles/stack-sunbird/templates/sunbird_nodebb.env [hunk 8]
```diff
@@ -1,10 +1,17 @@
...
+redis__password={{nodebb_redis_password}}
```
Comment (keshavprasadms) [inline ansible/roles/stack-sunbird/templates/sunbird_nodebb.env:13]: Same as above comment

File: ansible/roles/stack-sunbird/templates/sunbird_nodebb.env (hunk 6 above)
Comment (VenkatKadiveti) [reply ansible/roles/stack-sunbird/templates/sunbird_nodebb.env:7]: in load test env we face this heap memory issue with 1 GB after running soak test for 4hr with 45L samples 
that's why we use 2GB there so I am using same here (2GB)

Comment (VenkatKadiveti) [reply ansible/roles/stack-sunbird/templates/sunbird_nodebb.env]: replaced with 1GB

File: ansible/roles/stack-sunbird/templates/sunbird_nodebb.env (hunk 5 above)
Comment (VenkatKadiveti) [reply ansible/roles/stack-sunbird/templates/sunbird_nodebb.env]: Had a call with KP team member and he told me to use 4

**Review by keshavprasadms — APPROVED** [review]

**Pull request merged**